src/fomo.jl

```diff
@@ -0,0 +1,427 @@
+      max_iter = max_iter,
+      max_time = max_time,
+    ),
+  )
+
+  callback(nlp, solver, stats)
+
+  done = stats.status != :unknown
+
+  d .= ∇fk
+  norm_d = norm_∇fk
+  satβ = T(0)
+  ρk = T(0)
+  avgsatβ = T(0)
+  siter = 0
+  oneT = T(1)
+  while !done
+    λk = step_mult(αk,norm_d,backend)
+    c .= x .- λk .* d
+    uf = x == c # step addition underfow on every dimensions, should happen before αk == 0
+    if r2mode
+      ΔTk = norm_∇fk^2 * λk
+    else  
+      ΔTk = dot(∇fk , d) * λk
+    end
+    fck = obj(nlp, c)
+    if fck == -Inf
+      set_status!(stats, :unbounded)
+      break
+    end
+    ρk = (stats.objective - fck) / ΔTk
+    # Update regularization parameters
+    if ρk >= η2
+      αk = min(αmax, γ2 * αk)
+    elseif ρk < η1
+      αk = αk * γ1
+      if !r2mode
+        satβ *= γ3
+        d .= ∇fk .* (oneT - satβ) .+ m .* satβ
+      end
+    end
+
+    # Acceptance of the new candidate
+    if ρk >= η1
+      x .= c
+      if !r2mode
+        m .= ∇fk .* (oneT - β) .+ m .* β
+      end
+      set_objective!(stats, fck)
+      grad!(nlp, x, ∇fk)
+      norm_∇fk = norm(∇fk)
+      if !r2mode
+        satβ = find_beta(m, ∇fk, norm_∇fk, β, θ1, θ2)
+        d .= ∇fk .* (oneT - satβ) .+ m .* satβ
+        norm_d = norm(d)
+      else
+        d .= ∇fk
+        norm_d = norm_∇fk
+      end
+      if !r2mode
+        avgsatβ += satβ
+        siter += 1
+      end
+    end
+
+    set_iter!(stats, stats.iter + 1)
+    set_time!(stats, time() - start_time)
+    set_dual_residual!(stats, norm_∇fk)
+    optimal = norm_∇fk ≤ ϵ
+
+    if verbose > 0 && mod(stats.iter, verbose) == 0
+      @info infoline
+      if r2mode
+        infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk 1/αk
+      else
+        infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk αk satβ
+      end
+    end
+
+    set_status!(
+      stats,
+      get_status(
+        nlp,
+        elapsed_time = stats.elapsed_time,
+        optimal = optimal,
+        max_eval = max_eval,
+        iter = stats.iter,
+        max_iter = max_iter,
+        max_time = max_time,
+      ),
+    )
+    
+    callback(nlp, solver, stats)
+
+    uf      && set_status!(stats,:small_step)
+    αk == 0 && set_status!(stats,:exception) # :small_nlstep exception should happen before
+
+    done = stats.status != :unknown
+  end
+  if !r2mode
+    avgsatβ /= siter
+    stats.solver_specific[:avgsatβ] = avgsatβ
+  end
+  set_solution!(stats, x)
+  return stats
+end
+
+"""
+find_beta(m, ∇f, norm_∇f, β, θ1, θ2)
+
+Compute satβ which saturates the contibution of the momentum term to the gradient.
+`satβ` is computed such that the two gradient-related conditions are ensured: 
+1. [(1-satβ)∇f(xk) + satβ mk.∇f(xk)] ≥ θ1||∇f(xk)||²
+2. ||∇f(xk)|| ≥ θ2||(1-satβ)∇f(xk) + satβ mk.∇f(xk)||
+with `m` memory of past gradient/
+""" 
+function find_beta(m::V,∇f::V,norm_∇f::T, β::T, θ1::T, θ2::T) where {T,V}
```

Comment:
```suggestion
function find_beta(m::V, ∇f::V, norm_∇f::T, β::T, θ1::T, θ2::T) where {T,V}
```
Use proper spacing everywhere to make the code easier to read.